### PR TITLE
Feature/user linking

### PR DESCRIPTION
**Pull request by countcain**

### CODE REVIEW

**Comment by sebouh00** [comment]
@countcain Can you provide some context for this change?

**Comment by countcain** [comment]
> @countcain Can you provide some context for this change?

@sebouh00 Our client has issues when implementing user-linking through react native. This is for verifying the bridge and demonstrate how to use the user-linking feature.

File: ios/RNSentiance/AppDelegate.m [hunk 1]
```diff
@@ -23,10 +23,10 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
                                             initialProperties:nil];
 
   // Read SENTIANCE_APP_ID and SENTIANCE_APP_SECRET from any safe source
-  NSString * SENTIANCE_APP_ID = @"";
-  NSString * SENTIANCE_APP_SECRET = @"";
-  
-  [[bridge moduleForName:@"RNSentiance"] initSDK:SENTIANCE_APP_ID secret:SENTIANCE_APP_SECRET baseURL:nil shouldStart:YES resolver:nil rejecter:nil];
+  //NSString * SENTIANCE_APP_ID = @"";
```
[the remark this answers is not in the document]
Comment (countcain) [reply ios/RNSentiance/AppDelegate.m:26]: Nope. We should read a flag here checking whether the user linking is done on the JS side. If yes, we init the SDK on the native side. That being said on the JS side, once the user linking is finished, a flag should be set using [setValueForKey](https://github.com/sentiance/react-native-sentiance/blob/master/index.d.ts#L39)

File: App.js [hunk 2]
```diff
@@ -35,10 +37,24 @@ export default class App extends Component {
...
     this.sdkCrashEventSubscription.remove();
   }
 
-  async subscribe() {
+  async subscribeBeforeSDKSetup() {
+    this.userLinkListener = rnSentianceEmitter.addListener(
+      "SDKUserLink",
+      id => {
+        const { installId } = id;
+        setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (countcain) [reply App.js:49]: yes. This branch is still WIP. Will let you know once it is ready for review. :)

There is also a bug here while calling ```RNSentiance.userLinkCallback(bool);``` I am debugging it on this branch.

File: ios/RNSentiance/AppDelegate.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (sebouh00) [reply ios/RNSentiance/AppDelegate.m:26]: Understood. A couple of suggestions:
1. We should implement that logic as well for both iOS and Android.
2. With this change, the non-user-linking example will no longer be available. If we want to keep both, this change should either be left in its own branch, or be implemented in a forked repo (which is what we've done for native sample apps).

File: App.js [hunk 3]
```diff
@@ -51,42 +74,41 @@ export default class App extends Component {
...
+
+      await RNSentiance.initWithUserLinkingEnabled(
+        appId,
+        appSecret,
+        null,
+        true
+      );
     }
+
+    const userId = await RNSentiance.getUserId();
```
Comment (sebouh00) [inline App.js:99]: If init fails above, the `getUserId()` call seem to trigger `promise.reject()` in the RN plugin. What is assigned to `userId` then?

Comment (countcain) [reply App.js:99]: If `initWithUserLinkingEnabled` failed above, the error will be thrown immediately. The next tick of the process will be this line https://github.com/sentiance/react-native-sentiance-example/pull/6/files/480ffa3486ed228ce912b2ee3b58535740dfa74d#diff-958f6f56cd42c904603d226cd36a57b6R32

So RNSentiance.getUserId(); won't be invoked.

**Comment by countcain** [comment]
> `isSdkInitialized()` checks for the state "INITIALIZED". But when initialization is in progress, the state will be "INIT_IN_PROGRESS", and this will trigger a second init call, potentially resulting in a crash. It's better to use `isSdkNotInitialized()` which explicitly checks for "NOT_INITIALIZED" instead.

@sebouh00 Thanks for reminding this.

**Comment by countcain** [comment]
Close this one. The `user-linking` example will be tracked in the [user-linking](https://github.com/sentiance/react-native-sentiance-example/tree/user-linking) branch. [Diff from master](https://github.com/sentiance/react-native-sentiance-example/compare/sdk-update-and-refine-example...user-linking).

**Pull request closed without merging**